Panic on isValidTradePrice

Stacktrace

```sh
time="2021-06-23T13:11:32Z" level=debug msg=/Trade/TradePropose
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x927363]

goroutine 14604 [running]:
github.com/tdex-network/tdex-daemon/internal/core/application.isValidTradePrice(0x10b8680, 0xc01535a6c0, 0x0, 0xc014d1ce80, 0xc018e0aa80, 0x30, 0x48, 0xc00031e380, 0xfffffffe, 0x0)
	/tdex-daemon/internal/core/application/trade_service.go:1135 +0xa3
github.com/tdex-network/tdex-daemon/internal/core/application.(*tradeService).TradePropose(0xc000014300, 0x10b1480, 0xc013048bd0, 0xc016132d40, 0x40, 0xc016132d80, 0x40, 0x0, 0x10b8680, 0xc01535a6c0, ...)
	/tdex-daemon/internal/core/application/trade_service.go:369 +0x74a
github.com/tdex-network/tdex-daemon/internal/interfaces/grpc/handler.traderHandler.tradePropose(0x10b3878, 0xc000014300, 0xc01e7ea1c0, 0x10b7558, 0xc01cbf3a80, 0xdd1401, 0xc01cbf3a80)
	/tdex-daemon/internal/interfaces/grpc/handler/trade.go:229 +0x1cd
github.com/tdex-network/tdex-daemon/internal/interfaces/grpc/handler.traderHandler.TradePropose(...)
	/tdex-daemon/internal/interfaces/grpc/handler/trade.go:64
github.com/tdex-network/tdex-protobuf/generated/go/trade._Trade_TradePropose_Handler(0xdef9e0, 0xc0102a5b80, 0x10b5050, 0xc000274e40, 0xe42aa0, 0xc0003b05b0)
	/go/pkg/mod/github.com/tdex-network/tdex-protobuf@v0.0.0-20210507104156-d509331cccdb/generated/go/trade/trade_grpc.pb.go:282 +0x113
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1.1(0xdef9e0, 0xc0102a5b80, 0x10b5050, 0xc000274e40, 0x1, 0xc013048c00)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:71 +0xf1
github.com/tdex-network/tdex-daemon/internal/interfaces/grpc/interceptor.streamLogger(0xdef9e0, 0xc0102a5b80, 0x10b5050, 0xc000274e40, 0xc00038bd88, 0xc010dce5f0, 0x48, 0x7f62a5f8c108)
	/tdex-daemon/internal/interfaces/grpc/interceptor/logger.go:27 +0xbe
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1.1(0xdef9e0, 0xc0102a5b80, 0x10b5050, 0xc000274e40, 0xc0109c4f60, 0x0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:74 +0x99
github.com/tdex-network/tdex-daemon/internal/interfaces/grpc/interceptor.streamMacaroonAuthHandler.func1(0xdef9e0, 0xc0102a5b80, 0x10b5050, 0xc000274e40, 0xc00038bd88, 0xc010dce5f0, 0xc0125adc70, 0x41b9b8)
	/tdex-daemon/internal/interfaces/grpc/interceptor/auth.go:42 +0xd3
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1(0xdef9e0, 0xc0102a5b80, 0x10b5050, 0xc000274e40, 0xc00038bd88, 0xfda1c8, 0x10b1480, 0xc013048bd0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:79 +0x155
google.golang.org/grpc.(*Server).processStreamingRPC(0xc010358540, 0x10b90b8, 0xc015304d80, 0xc01cc30d80, 0xc0109c51d0, 0x1516e20, 0x0, 0x0, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:1540 +0x535
google.golang.org/grpc.(*Server).handleStream(0xc010358540, 0x10b90b8, 0xc015304d80, 0xc01cc30d80, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:1613 +0xca5
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc011e805d0, 0xc010358540, 0x10b90b8, 0xc015304d80, 0xc01cc30d80)
	/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:934 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:932 +0x1fd

```﻿

--- a/internal/core/application/trade_service.go
+++ b/internal/core/application/trade_service.go
@@ -1117,22 +1117,21 @@ func isValidTradePrice(
 	unspents []domain.Unspent,
 	slippage decimal.Decimal,
 ) bool {
-	// TODO: parallelize the 2 ways of calculating and validating the preview
-	// amount to speed up the process.
 	amount := swapRequest.GetAmountR()
 	if tradeType == TradeSell {
 		amount = swapRequest.GetAmountP()
 	}
 
-	preview, _ := previewForMarket(
+	preview, err := previewForMarket(
 		unspents,
 		market,
 		tradeType,
 		amount,
 		market.BaseAsset,
 	)
-
-	if preview != nil {
+	if err != nil {
+		log.Debugf("preview failed for reason: %s", err)
+	} else {
 		if isPriceInRange(swapRequest, tradeType, preview.amount, true, slippage) {
 			return true
 		}
@@ -1143,7 +1142,7 @@ func isValidTradePrice(
 		amount = swapRequest.GetAmountR()
 	}
 
-	preview, _ = previewForMarket(
+	preview, err = previewForMarket(
 		unspents,
 		market,
 		tradeType,
@@ -1151,7 +1150,8 @@ func isValidTradePrice(
 		market.QuoteAsset,
 	)
 
-	if preview == nil {
+	if err != nil {
+		log.Debugf("preview failed for reason: %s", err)
 		return false
 	}
 

--- a/internal/core/application/trade_service.go
+++ b/internal/core/application/trade_service.go
@@ -1132,8 +1132,10 @@ func isValidTradePrice(
 		market.BaseAsset,
 	)
 
-	if isPriceInRange(swapRequest, tradeType, preview.amount, true, slippage) {
-		return true
+	if preview != nil {
+		if isPriceInRange(swapRequest, tradeType, preview.amount, true, slippage) {
+			return true
+		}
 	}
 
 	amount = swapRequest.GetAmountP()
@@ -1148,6 +1150,10 @@ func isValidTradePrice(
 		amount,
 		market.QuoteAsset,
 	)
+
+	if preview == nil {
+		return false
+	}
 
 	return isPriceInRange(swapRequest, tradeType, preview.amount, false, slippage)
 }

--- a/internal/core/application/trade_service.go
+++ b/internal/core/application/trade_service.go
@@ -1117,21 +1117,22 @@ func isValidTradePrice(
 	unspents []domain.Unspent,
 	slippage decimal.Decimal,
 ) bool {
+	// TODO: parallelize the 2 ways of calculating and validating the preview
+	// amount to speed up the process.
 	amount := swapRequest.GetAmountR()
 	if tradeType == TradeSell {
 		amount = swapRequest.GetAmountP()
 	}
 
-	preview, err := previewForMarket(
+	preview, _ := previewForMarket(
 		unspents,
 		market,
 		tradeType,
 		amount,
 		market.BaseAsset,
 	)
-	if err != nil {
-		log.Debugf("preview failed for reason: %s", err)
-	} else {
+
+	if preview != nil {
 		if isPriceInRange(swapRequest, tradeType, preview.amount, true, slippage) {
 			return true
 		}
@@ -1142,7 +1143,7 @@ func isValidTradePrice(
 		amount = swapRequest.GetAmountR()
 	}
 
-	preview, err = previewForMarket(
+	preview, _ = previewForMarket(
 		unspents,
 		market,
 		tradeType,
@@ -1150,8 +1151,7 @@ func isValidTradePrice(
 		market.QuoteAsset,
 	)
 
-	if err != nil {
-		log.Debugf("preview failed for reason: %s", err)
+	if preview == nil {
 		return false
 	}
 

